src/app/utils/xml-utils.ts

```diff
@@ -36,3 +36,74 @@ export function replaceNotWordChar(textContent: string) {
 export function removeSpaces(textContent: string) {
   return textContent.replace(/\s/g, '');
 }
+
+/**
+ * It removes excessive spaces, any tabulation, new lines and non-word characters
+ */
+export function normalizeSpaces(textContent: string) {
+  return textContent.replace(/[\s]{2,}|\n|\t|\r/g, ' ').trimStart().trimEnd();
+}
+
+/**
+* Significant text sometimes is split inside two or more text evt-element inside the main one, especially when it contains new line characters.
+* This function returns a string with all the text elements chained
+*/
+export function chainFirstChildTexts(elem: XMLElement, evtTextComplexElements: string[], evtInnerTextElements: string[]): string {
+  if (elem === undefined) { return ''; };
+  const evtTextElements = {
+    '#text': 'nodeValue',
+    'p': 'textContent',
+  };
+  let out = '';
+  elem.childNodes.forEach((node) => (evtTextElements[node.nodeName] !== undefined) ? out += node[ evtTextElements[node.nodeName] ] : (
+    evtTextComplexElements.includes(node.nodeName) ? out += chainDeepTexts(node, evtInnerTextElements) : '' ))
+
+  return out;
+}
+
+/**
+ * Retrieve and chain textContent of all descendents
+ * @param elem ChildNode
+ * @returns out string
+ */
+export function chainDeepTexts(elem: ChildNode, evtInnerTextElements: string[]): string {
+  const textProperty = 'textContent';
+  let out = '';
+  elem.childNodes.forEach((x) => (evtInnerTextElements.includes(x.nodeName)) ? ((x[textProperty] !== null) ? out += x[textProperty] : '') : '')
+
+  return out;
+}
+
+/**
+* Retrieve external elements searching between provided elements types and filtering on attributes base
+* It searches all document for a bibl with the correct xml:id.
+* It would be faster if we knew the id or unique element to search in
+*/
+export function getExternalElements(elem: XMLElement, attrSourceNames: string[], attrTargetName: string, elTypes: string): XMLElement[] {
```

Comment:
Wha does it mean "external elements"?

This comment "It searches all document for a bibl with the correct xml:id" suggest me that this is a specific function for bibl?

Reply:
It's here because it's used in 2-3 elements (analogue, sources and bibl), I think I've also used it in the next feature I will soon submit to PR.
Where would you place it?